language/vm/vm-runtime/src/libra_vm.rs

```diff
@@ -1,100 +1,554 @@
+        let signature_verified_block: Vec<Result<SignatureCheckedTransaction, VMStatus>> =
+            txn_block
+                .into_par_iter()
+                .map(|txn| {
+                    txn.check_signature()
+                        .map_err(|_| VMStatus::new(StatusCode::INVALID_SIGNATURE))
+                })
+                .collect();
+        let mut result = vec![];
+        for transaction in signature_verified_block {
+            record_stats! {time_hist | TXN_TOTAL_TIME_TAKEN | {
+                    let output = match transaction {
+                        Ok(txn) => self.execute_user_transaction(state_view, data_cache, &txn),
+                        Err(e) => discard_error_output(e),
+                    };
+                    report_execution_status(output.status());
 
+                    // `result` is initially empty, a single element is pushed per loop iteration and
+                    // the number of iterations is bound to the max size of `signature_verified_block`
+                    assume!(result.len() < usize::max_value());
+                    result.push(output);
+                }
+            }
+        }
         Ok(result)
     }
 }
 
+pub(crate) fn discard_error_output(err: VMStatus) -> TransactionOutput {
+    // Since this transaction will be discarded, no writeset will be included.
+    TransactionOutput::new(
+        WriteSet::default(),
+        vec![],
+        0,
+        TransactionStatus::Discard(err),
+    )
+}
+
+// Validators external API
+impl VMVerifier for LibraVM {
+    /// Determine if a transaction is valid. Will return `None` if the transaction is accepted,
+    /// `Some(Err)` if the VM rejects it, with `Err` as an error code. We verify the following
+    /// items:
+    /// 1. The signature on the `SignedTransaction` matches the public key included in the
+    ///    transaction
+    /// 2. The script to be executed is in the whitelist.
+    /// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
+    /// right sequence number and the sender has enough balance to pay for the gas.
+    /// 4. Transaction arguments matches the main function's type signature.
+    fn validate_transaction(
+        &self,
+        transaction: SignedTransaction,
+        state_view: &dyn StateView,
+    ) -> Option<VMStatus> {
+        let data_cache = BlockDataCache::new(state_view);
+        record_stats! {time_hist | TXN_VALIDATION_TIME_TAKEN | {
+                let gas_schedule = self.move_vm.load_gas_schedule(state_view, &data_cache);
+                let signature_verified_txn = match transaction.check_signature() {
+                    Ok(t) => t,
+                    Err(_) => return Some(VMStatus::new(StatusCode::INVALID_SIGNATURE)),
+                };
+                let res = match self.verify_transaction_impl(&signature_verified_txn, gas_schedule.as_ref().map_err(|err| err.clone()), state_view, &data_cache) {
```

Comment:
Once again, this is a pretty ugly hack for cost table. The main reason is that processing `WriteSet` txn should not require a gas table, yet this logic is combined with processing regular user transactions for now.